File: Assets/ToLua/Lua/user/ui/hero/model/hero_info.lua
local role_info = require('ui/role/model/role_info')
local hero_info = role_info.extend()

hero_info.__index = hero_info

function hero_info:New( instanceID,  heroDataID,  breakthroughLevel,  strengthenLevel,  advanceLevel,  level)
  local o = {}
  setmetatable(o, hero_info)
  o:Init()
  o:Set(instanceID,  heroDataID,  breakthroughLevel,  strengthenLevel,  advanceLevel,  level)
  return o
end
function hero_info:NewByGameResData (gameResData)
  local instanceID = 0
  local heroDataID = gameResData.id
  local breakthroughLevel = 0
  local strengthenLevel = 0
  local advanceLevel = gameResData.star
  local level = 1
  
  return hero_info:New(instanceID,  heroDataID,  breakthroughLevel,  strengthenLevel,  advanceLevel,  level)
end
function hero_info:NewByHeroProtoData(heroProtoData)
  local o = {}
  setmetatable(o, hero_info)
  o:Init()
  o:SetHeroInfo(heroProtoData)
  return o
end
function hero_info.NewByDrawCardDropProto(drawDropCardProto)
  local o = {}
  setmetatable(o, hero_info)
  o:Init()
  o:Set(drawDropCardProto.id,drawDropCardProto.no,0,0,drawDropCardProto.star,1)
  return o
end

function hero_info:NewByTeamHeroProtoData(data)
  local o = {}
  setmetatable(o, hero_info)
  o:Init()
  if data.lv ~= 1 then
    self.level = data.lv
    
  end
  if data.star ~= 0 then
    self.advanceLevel = data.star
  end
  if data.aggrLv ~= -1 then
    self.strengthenLevel = data.aggrLv
  end
  if data.breakLayer ~= 0 then
    self.breakthroughLevel = data.breakLayer
  end
  return o
end

function hero_info:Clone()
  return self:GetHeroInfoCopy()
end

function hero_info:GetHeroInfoCopy()
  local info = hero_info:New(self.instanceID, self.heroData.id,  self.breakthroughLevel,self.strengthenLevel,self.advanceLevel,self.level)
  info.exp = self.exp
  info.strengthenExp = self.strengthenExp
  if self.relations then
    info.relations = {}
    for k,v in ipairs(self.relations) do
      info.relations[k] = v
    end
  end
  return info
end

function hero_info:Init()
  self.instanceID = 0
  self.modelId = 0
  self.breakLayer = 1
  self.star = 0
  self.exp = 0
  self.level = 1
  self.strengthenLevel = 0
  self.strengthenExp = 0
  self.relations = {}
  self.myPower = 0
  self.needUpdatePower = true
  self.isLocked = false
end

function hero_info:Set(instanceID,  heroDataID,  breakthroughLevel,  strengthenLevel,  advanceLevel,  level)
  self.instanceID = instanceID
  local hd = gamemanager.GetData('hero_data')  
  self.heroData = hd.GetDataById(heroDataID)
  if(self.heroData == nil) then
    print("these is no heroId:"..heroDataID)
  end
  self.breakthroughLevel = breakthroughLevel
  self.strengthenLevel = strengthenLevel
  self.advanceLevel = advanceLevel
  self.level = level
  self.needUpdatePower = true
end

function hero_info:SetHeroInfo(heroProtoData)
  self.instanceID = heroProtoData.id
  self.modelId=heroProtoData.modelId
  if heroProtoData.modelId ~= 0 then
    local hd = gamemanager.GetData('hero_data')
    self.heroData = hd.GetDataById(heroProtoData.modelId)
    if(self.heroData == nil) then
      print("these is no heroId:"..heroProtoData.modelId)
    end
  end
  if heroProtoData.breakLayer ~= 0 then 
    self.breakthroughLevel = heroProtoData.breakLayer
  end
  
  if heroProtoData.star ~= 0 then 
    self.advanceLevel = heroProtoData.star
  end
  
  if heroProtoData.exp ~= -1 then
    self.exp = heroProtoData.exp
  end
  if heroProtoData.lv ~= 0 then 
    self.level = heroProtoData.lv
  end
  --print('update hero strengthen','lv:'..self.strengthenLevel,'exp:'..self.strengthenExp,'resp.lv:'..heroProtoData.aggrLv,'resp.exp:'..heroProtoData.aggrExp)
  
  if heroProtoData.aggrLv ~= -1 then
    self.strengthenLevel = heroProtoData.aggrLv
  end
  
  if heroProtoData.aggrExp ~= -1 then
    self.strengthenExp = heroProtoData.aggrExp
  end
  --HeroRelationProto
  if heroProtoData.isUpRelations then  
    self.relations = {}
    for k,v in ipairs(heroProtoData.relations) do
      self.relations[k] = v
    end
  end
  self.needUpdatePower = true
end

function hero_info:IsActiveRelation(relationId)
  for k,v in ipairs(self.relations) do
    if v.id == relationId then
      return true
    end
  end
  return false
end
--英雄自身战力（不包括穿的装备）
function hero_info:GetMyPower()
  if self.needUpdatePower then
    self.needUpdatePower = false
    self.myPower = self:Power()
  end
  return self.myPower
end

return hero_info

File: Assets/Code/Logic/Hero/Model/HeroInfo.cs
using UnityEngine;
using System.Collections.Generic;
using Logic.Protocol.Model;
using Logic.Enums;
using Logic.Equipment.Model;
using Logic.Role.Model;
using LuaInterface;
using Logic.Game.Model;

namespace Logic.Hero.Model
{
    public class HeroInfo : RoleInfo
    {
        //public uint instanceID;
        //public HeroData heroData;
        public List<RoleAttribute> heroAttributes;

        //public int breakthroughLevel = 0;
        //public int strengthenLevel = 0;
        //public int advanceLevel = 0;
        //public int level = 1;

        //public int exp = 0;
        //public int strengthenExp = 0;
        //public int weaponID = 0;
        //public int armorID = 0;
        //public int accessoryID = 0;  

        public HeroInfo(uint instanceID, int heroDataID, int breakthroughLevel, int strengthenLevel, int advanceLevel, int level = 1)
        {
            this.instanceID = instanceID;
            this.heroData = HeroData.GetHeroDataByID(heroDataID);
            if (this.heroData == null)
                Debugger.LogError("heroData is null,can't find id :" + heroDataID);
            this.breakthroughLevel = breakthroughLevel;
            this.strengthenLevel = strengthenLevel;
            this.advanceLevel = advanceLevel;
            this.level = level;
            UpdateHeroAttribute();

        }

        public HeroInfo GetHeroInfoCopy()
        {
            HeroInfo heroInfoCopy = new HeroInfo(this.instanceID, this.heroData.id, this.breakthroughLevel, this.strengthenLevel, this.advanceLevel, this.level);
            return heroInfoCopy;
        }

        public HeroInfo(HeroProtoData heroProtoData)
        {
            SetHeroInfo(heroProtoData);
        }

		public HeroInfo (LuaTable heroInfoLuaTable)
		{
			this.instanceID = heroInfoLuaTable["instanceID"].ToString().ToUInt32();
			LuaTable heroDataLuaTable = (LuaTable)heroInfoLuaTable["heroData"];
			if(heroDataLuaTable == null)
				Debugger.LogError("heroData is null,can't find ");
			this.heroData = HeroData.GetHeroDataByID(heroDataLuaTable["id"].ToString().ToInt32());
			this.breakthroughLevel = heroInfoLuaTable["breakthroughLevel"].ToString().ToInt32();
			this.strengthenLevel = heroInfoLuaTable["strengthenLevel"].ToString().ToInt32();
			this.advanceLevel = heroInfoLuaTable["advanceLevel"].ToString().ToInt32();
			this.level = heroInfoLuaTable["level"].ToString().ToInt32();

			this.exp = heroInfoLuaTable["exp"].ToString().ToInt32();
			this.strengthenExp = heroInfoLuaTable["strengthenExp"].ToString().ToInt32();

			this.weaponID = heroInfoLuaTable["weaponID"].ToString().ToInt32();
			this.armorID = heroInfoLuaTable["armorID"].ToString().ToInt32();
			this.accessoryID = heroInfoLuaTable["accessoryID"].ToString().ToInt32();
		}

        public HeroInfo(DrawCardDropProto drawDropCardProto)
        {
            this.instanceID = 0;
            this.heroData = HeroData.GetHeroDataByID(drawDropCardProto.no);
            this.breakthroughLevel = 0;
            this.strengthenLevel = 0;
            this.advanceLevel = drawDropCardProto.star;
            this.level = 1;
        }
        public HeroInfo(TeamHeroProtoData data)
        {
            this.instanceID = (uint)data.id;
            heroData = HeroData.GetHeroDataByID(data.heroNo);
            if (data.lv != 1)
                level = data.lv;
            if (data.star != 0)
                advanceLevel = data.star;
            if (data.aggrLv != -1)
                strengthenLevel = data.aggrLv;
            if (data.breakLayer != 0)
                breakthroughLevel = data.breakLayer;
        }
        //		public HeroInfo (BuyHeroInfo buyHeroInfo)
        //		{
        //			this.instanceID = 0;
        //			this.heroData = HeroData.GetHeroDataByID(buyHeroInfo.modelId);
        //			this.advanceLevel = buyHeroInfo.star;
        //			UpdateHeroAttribute();
        //		}

        public HeroInfo(int heroId)
        {
            this.instanceID = (uint)heroId;
            heroData = HeroData.GetHeroDataByID(heroId);
            this.advanceLevel = 1;
            UpdateHeroAttribute();
        }

        public HeroInfo(HeroData heroData)
        {
            this.heroData = heroData;
            this.advanceLevel = 1;//召唤兽默认用1阶模型
            UpdateHeroAttribute();
        }

		public HeroInfo (GameResData gameResData)
		{
			this.heroData = HeroData.GetHeroDataByID(gameResData.id);
			this.advanceLevel = gameResData.star;
			UpdateHeroAttribute();
		}

        public void SetHeroInfo(HeroProtoData heroProtoData)
        {
            this.instanceID = (uint)heroProtoData.id;
            if (heroProtoData.modelId != 0)
                this.heroData = HeroData.GetHeroDataByID(heroProtoData.modelId);
            if (heroProtoData.breakLayer != 0)
            {
                this.breakthroughLevel = heroProtoData.breakLayer;
            }
            if (heroProtoData.aggrLv != -1)
            {
                this.strengthenLevel = heroProtoData.aggrLv;
            }
            if (heroProtoData.star != 0)
            {
                this.advanceLevel = heroProtoData.star;
            }
            if (heroProtoData.exp != -1)
            {
                this.exp = heroProtoData.exp;
            }
            if (heroProtoData.lv != 0)
            {
                this.level = heroProtoData.lv;
            }
            if (heroProtoData.aggrExp != -1)
            {
                this.strengthenExp = heroProtoData.aggrExp;
            }
            UpdateHeroAttribute();

        }

        private void UpdateHeroAttribute()
        {
            heroAttributes = HeroUtil.CalcHeroAttributes(this);
        }
        public override int modelDataId
        {
            get
            {
                return heroData.id;
            }
        }

        public override int MaxAdvanceLevel
        {
            get
            {
                return (int)heroData.starMax;
            }
        }

    }
}